TME-117 - New RAG on how to extend CF access native features with ser… (#17317)
cloudflare/cloudflare-docs: src/content/docs/reference-architecture/diagrams/sase/augment-access-with-serverless.mdx

---
title: Extend ZTNA with external authorization and serverless computing
pcx_content_type: reference-architecture-diagram
products:
  - Access
  - Workers
sidebar:
  order: 1
  label: ZTNA with external authorization
updated: 2024-10-10
---

import { RelatedProduct, LinkCard} from "~/components";

## Introduction

> Companies using Zero Trust Network Access (ZTNA) services build policies to determine if a user can access a protected resource such as a privately hosted Wiki server or source code repository. Policies typically use group membership, authentication methods, device security posture to determine which users can access which resources. 

Reviewer: ```suggestion
Companies use Zero Trust Network Access (ZTNA) services to build policies to determine if a user can access protected resources such as privately hosted wiki servers or source code repositories. Policies typically use group membership, authentication methods or device security posture to determine which users can access which resources. 
```

Secure access requires a range of attributes being available to the policy engine for evaluation. With Cloudflare's ZTNA service, [Access](/cloudflare-one/policies/access/), it is possible to include in the policy an external request to another API that provides part of the data required for the access decision. 

> For example, you might have a policy which states all members of the group "Engineers", who have authenticated with credentials that required a hard token, can have access to the self-hosted source code repository. But you also want to only allow engineers who have completed security training. That data might be available in another system, so Cloudflare allows you to, as part of the policy check, make a call using [Workers](https://workers.cloudflare.com/) to the training system to determine if this user has passed security training.

Reviewer: ```suggestion
For example, you might have a policy which states all members of the group "Engineers", who have authenticated with credentials that required a hard token, can have access to the self-hosted source code repository. But you also want to only allow engineers who have completed security training. That data might be available in another system, so as part of the policy check, Cloudflare allows you to make a call to the training system using [Workers](https://workers.cloudflare.com/) to determine if a user has passed security training.
```

To help integrate these types of use cases, Cloudflare has an [entire development platform](/workers/) on which you can design and run your own business logic. This means you spend less time trying to piece a solution together and more time getting the integration done.

This document outlines how to combine both solutions to enhance Cloudflare Access capabilities in terms of [authorization and authentication](https://www.cloudflare.com/learning/access-management/what-is-access-control/).

## Showcased products

<RelatedProduct header="Workers" href="/workers/" product="workers">
    Build serverless applications and deploy instantly across the globe for exceptional performance, reliability, and scale.
</RelatedProduct>

<RelatedProduct header="Access" href="/cloudflare-one/" product="cloudflare-one">

>     Cloudflare Zero Trust replaces legacy security perimeters with Cloudflare's global network, making the Internet faster and safer for teams around the world

Reviewer: ```suggestion
    Cloudflare Zero Trust replaces legacy security perimeters with Cloudflare's global network, making the Internet faster and safer for teams around the world.
```

    href="/cloudflare-one/policies/access/external-evaluation/"
/>

### 2. Analyze and validate the authentication material (JWT)

When a user successfully authenticates and is authorized to access a protected application, Cloudflare inserts a [JSON Web Token (JWT)](/cloudflare-one/identity/authorization-cookie/validating-json/) into the HTTP traffic sent to the origin. This token serves as a valuable asset for expanding custom business logic through secure processing. The format for that JWT is deterministic and rather lightweight to avoid overloading the requests towards origin unnecessarily. 

Here is an example of a JWT sent to an origin (use [JWT.io](http://jwt.io) to read the contents of a JWT)

```json title="JWT content"
{
  "aud": [
    "264063895705477af73bfbaed1bf401981f4812eefcdb9fea33f5e10e666e282"
  ],
  "email": "john.doe@cloudflare.com",
  "exp": 1728551137,
  "iat": 1728464737,
  "nbf": 1728464737,
  "iss": "https://myorg.cloudflareaccess.com",
  "type": "app",
  "identity_nonce": "IA0hPRvwILtbUXSQ",
  "sub": "ce40d564-c72f-475f-a9b8-f395f19ad986",
  "device_id": "8469d7c4-83a9-11ee-b559-76e6e80876db",
  "country": "FR"
}
```

> Cloudflare exposes a specific [endpoint](/cloudflare-one/identity/authorization-cookie/validating-json/#_top) to allow anyone to validate and expand a Cloudflare signed JWT.

Reviewer 2: @nouvellonsteph , resolving this comment will give you the ✅!

As mentioned above, it'd be helpful to email support @ hyperlint. I've got a quick question or two that's just not relevant for this PR.

Please ping me directly on GH with any other questions or concerns!

> Cloudflare exposes a specific [endpoint](/cloudflare-one/identity/authorization-cookie/validating-json/#_top) to allow anyone to validate and expand a Cloudflare signed JWT.

Reviewer: @bllchmbrs Thanks so much for reaching out about this! I'm the PM for the PCX team and I'll email you shortly to see if I can get you the info you're after.
